src/realm_dart_sync.cpp

```diff
@@ -0,0 +1,131 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////////
+
+#include <realm/object-store/sync/sync_session.hpp>
+#include <realm/sync/config.hpp>
+#include "realm_dart.hpp"
+#include "realm_dart_sync.h"
+
+RLM_API void realm_dart_sync_client_log_callback(realm_userdata_t userdata, realm_log_level_e level, const char* message)
+{
+    auto ud = reinterpret_cast<realm_dart_userdata_async_t>(userdata);
+    ud->scheduler->invoke([ud, level, message=std::string(message)]() {
+        (reinterpret_cast<realm_log_func_t>(ud->dart_callback))(ud->handle, level, message.c_str());
+    });
+}
+
+RLM_API void realm_dart_sync_error_handler_callback(realm_userdata_t userdata, realm_sync_session_t* session, realm_sync_error_t error)
+{
+    // the pointers in error are to stack values, we need to make copies and move them into the scheduler invocation
+    struct error_copy {
+        std::string message;
+        std::string detailed_message;
+        std::map<std::string, std::string> user_info_map;
+        std::vector<realm_sync_error_user_info_t> user_info_vector;
+    } buf;
+
+    buf.message = error.error_code.message;
+    error.error_code.message = buf.message.c_str();
+
+    buf.detailed_message = error.detailed_message;
+    error.detailed_message = buf.detailed_message.c_str();
+
+    buf.user_info_vector.reserve(error.user_info_length);
+    for (size_t i = 0; i < error.user_info_length; i++) {
+        auto [it, _] = buf.user_info_map.emplace(error.user_info_map[i].key, error.user_info_map[i].value);
+        buf.user_info_vector[i].key = it->first.c_str();
+        buf.user_info_vector[i].value = it->second.c_str();
+    }
+    error.user_info_map = buf.user_info_vector.data();
+
+    auto ud = reinterpret_cast<realm_dart_userdata_async_t>(userdata);
+    ud->scheduler->invoke([ud, session=*session, error=std::move(error), buf=std::move(buf)]() {
+        (reinterpret_cast<realm_sync_error_handler_func_t>(ud->dart_callback))(ud->handle, const_cast<realm_sync_session_t*>(&session), error);
+    });
+}
+
+RLM_API void realm_dart_sync_wait_for_completion_callback(realm_userdata_t userdata, realm_sync_error_code_t* error)
+{
+    // we need to make a deep copy of error, because the message pointer points to stack memory
+    struct realm_dart_sync_error_code : realm_sync_error_code
+    {
+        realm_dart_sync_error_code(const realm_sync_error_code& error)
+        : realm_sync_error_code(error)
+        , message_buffer(error.message)
+        {
+            message = message_buffer.c_str();
+        }
+
+        const std::string message_buffer;
+    };
+
+    std::unique_ptr<realm_dart_sync_error_code> error_copy;
```

Comment:
So the copies lifetime is managed here. I like it, but we need to make sure that we are not trying to deallocate the string on the dart side, which we did before